Makefile

```diff
@@ -0,0 +1,164 @@
+# We cannot use $(shell pwd), which will return unix path format on Windows,
+# making it hard to use.
+cur_dir = $(dir $(abspath $(firstword $(MAKEFILE_LIST))))
+
+TOP := $(cur_dir)
+# RUSTFLAGS that are likely to be tweaked by developers. For example,
+# while we enable debug logs by default here, some might want to strip them
+# for minimal code size / consumed cycles.
+CUSTOM_RUSTFLAGS := -C debug-assertions
+# Additional cargo args to append here. For example, one can use
+# make test CARGO_ARGS="-- --nocapture" so as to inspect data emitted to
+# stdout in unit tests
+CARGO_ARGS :=
+MODE := release
+# Tweak this to change the clang version to use for building C code. By default
+# we use a bash script with somes heuristics to find clang in current system.
+CLANG := $(shell $(TOP)/scripts/find_clang)
+CLANG_OPTIMIZER := -O2
+# When this is set, a single contract will be built instead of all contracts
+CONTRACT :=
+# By default, we would clean build/{release,debug} folder first, in case old
+# contracts are mixed together with new ones, if for some reason you want to
+# revert this behavior, you can change this to anything other than true
+CLEAN_BUILD_DIR_FIRST := true
+BUILD_DIR := build/$(MODE)
+
+ifeq (release,$(MODE))
+	MODE_ARGS := --release
+endif
+
+# Pass setups to child make processes
+export CUSTOM_RUSTFLAGS
+export TOP
+export CARGO_ARGS
+export MODE
+export CLANG
+export CLANG_OPTIMIZER
+export BUILD_DIR
+
+default: build test
+
+build:
+	@if [ "x$(CLEAN_BUILD_DIR_FIRST)" = "xtrue" ]; then \
+		echo "Cleaning $(BUILD_DIR) directory..."; \
+		rm -rf $(BUILD_DIR); \
+	fi
+	mkdir -p $(BUILD_DIR)
+	@set -eu; \
+	if [ "x$(CONTRACT)" = "x" ]; then \
+		for contract in $(wildcard contracts/*); do \
+			$(MAKE) -e -C $$contract build; \
+		done; \
+		for crate in $(wildcard crates/*); do \
+			cargo build -p $$(basename $$crate) $(MODE_ARGS) $(CARGO_ARGS); \
+		done; \
+		for sim in $(wildcard native-simulators/*); do \
+			cargo build -p $$(basename $$sim) $(CARGO_ARGS); \
+		done; \
+	else \
+		$(MAKE) -e -C contracts/$(CONTRACT) build; \
+		cargo build -p $(CONTRACT)-sim; \
+	fi;
+
+report:
+	for contract in $(wildcard contracts/*); do \
+		du -h build/$(MODE)/$$(basename $$contract); \
+		ckb-debugger --bin build/$(MODE)/$$(basename $$contract); \
+	done; \
+
+# Run a single make task for a specific contract. For example:
+#
+# make run CONTRACT=stack-reorder TASK=adjust_stack_size STACK_SIZE=0x200000
+TASK :=
+run:
+	$(MAKE) -e -C contracts/$(CONTRACT) $(TASK)
+
+# test, check, clippy and fmt here are provided for completeness,
+# there is nothing wrong invoking cargo directly instead of make.
+test:
+	cargo test $(CARGO_ARGS)
+
+check:
+	cargo check $(CARGO_ARGS)
+
+clippy:
+	cargo clippy $(CARGO_ARGS)
+
+NON_RUST_CONTRACTS := contracts/secp256k1_ecdsa contracts/secp256k1_schnorr
```

Comment:
Personally, I would favor any of the following choices:

* Use separate folders for Rust and C contracts. For example, Rust contracts live in `contracts`, while C contracts can live in `c` or `c-contracts`
* All C contracts must use a common prefix such as `c-`
* Add a `fmt` task at each Rust contract level, it could be a dummy command that does nothing, or it could run `cargo fmt`, both work.

Right now it relies on manual maintenance to keep track of all C contracts, I'm not sure if this is a good idea.

Reply:
> Personally, I would favor any of the following choices:
> 
> * Use separate folders for Rust and C contracts. For example, Rust contracts live in `contracts`, while C contracts can live in `c` or `c-contracts`
> * All C contracts must use a common prefix such as `c-`
> * Add a `fmt` task at each Rust contract level, it could be a dummy command that does nothing, or it could run `cargo fmt`, both work.
> 
> Right now it relies on manual maintenance to keep track of all C contracts, I'm not sure if this is a good idea.

I think the third option is probably the best, as it has the best scalability. Currently each rust contract already contains fmt task, so there is no need to add it.